src/app/src/components/EmbeddedMapConfigWrapper.jsx

```diff
@@ -0,0 +1,130 @@
+import apiRequest from '../util/apiRequest';
+import EmbeddedMapConfig from './EmbeddedMapConfig';
+import {
+    formatExistingConfig,
+    formatExistingFields,
+    combineEmbedAndNonstandardFields,
+    formatEmbedConfigForServer,
+    getErrorMessage,
+} from '../util/embeddedMap';
+
+function EmbeddedMapConfigWrapper({ user }) {
+    const [loading, setLoading] = useState(true);
+    const [errors, setError] = useState(null);
+    const [embedConfig, setEmbedConfig] = useState(
+        formatExistingConfig(user.embed_config),
+    );
+    const [embedFields, setEmbedFields] = useState(
+        formatExistingFields(embedConfig.embed_fields),
+    );
+
+    const handleError = e => {
+        setError(getErrorMessage(e));
+        setLoading(false);
+        setTimeout(() => setError(null), 5000);
+    };
+
+    const fetchFields = useCallback(() => {
+        setLoading(true);
+        apiRequest
+            .get(makeNonStandardFieldsURL())
+            .then(({ data }) => {
+                setEmbedFields(
+                    combineEmbedAndNonstandardFields(
+                        user.embed_config.embed_fields,
+                        data,
+                    ),
+                );
+                setLoading(false);
+            })
+            .catch(e => handleError(e));
+    }, [user]);
+
+    const createConfig = () => {
+        setLoading(true);
+        apiRequest
+            .post(
+                makeEmbedConfigURL(),
+                formatEmbedConfigForServer(embedConfig, embedFields),
+            )
+            .then(({ data }) => {
+                setEmbedConfig(formatExistingConfig(data));
+                setLoading(false);
+            })
+            .catch(e => handleError(e));
+    };
+
+    const updateConfig = () => {
+        setLoading(true);
+        apiRequest
+            .put(
+                makeEmbedConfigURL(embedConfig.id),
+                formatEmbedConfigForServer(embedConfig, embedFields),
+            )
+            .then(({ data }) => {
+                setEmbedConfig(formatExistingConfig(data));
+                setLoading(false);
+            })
+            .catch(e => handleError(e));
+    };
+
+    // Fetches and merges nonstandard fields from the database with fields
+    // from the user's EmbedConfig
+    useEffect(() => fetchFields(), [user, fetchFields]);
+
+    /* eslint-disable react-hooks/exhaustive-deps */
+    // Disabled exhaustive dependencies warning because we don't want to rerun
+    // in response to the loading state changing
+    // When the embed cofig or fields are updated, wait for 500ms. If no
+    // additional changes are made, update/create the embed config in the database.
+    useEffect(() => {
+        const handler = setTimeout(() => {
+            if (loading) return;
+            if (embedConfig.id) {
+                updateConfig();
+            } else {
+                createConfig();
+            }
+        }, 500);
```

Comment:
Did we choose 500ms for any specific reason?

I am seeing "polling" like behavior where a PUT requests are being made repeatedly. I was expecting that PUT requests would only be made in response to a field changing.

![2021-05-11 14 41 23](https://user-images.githubusercontent.com/17363/117888564-1cd48f00-b267-11eb-86bd-99a8537be29c.gif)

Reply:
The polling behavior was caused because the update and create functions are being recreated on each render. Removing them from the dependency array restored expected behavior. The 500ms delay is meant to prevent repeated calls to the database while the user is actively typing, but the actual duration of the debounce is fairly random.